rust/lance-index/src/scalar/btree.rs

```diff
@@ -0,0 +1,809 @@
+            .column(2)
+            .as_any()
+            .downcast_ref::<UInt32Array>()
+            .unwrap();
+        let offsets = data
+            .column(3)
+            .as_any()
+            .downcast_ref::<UInt64Array>()
+            .unwrap();
+
+        let first_min = ScalarValue::try_from_array(&mins, 0)?;
+        map.entry(OrderableScalarValue(first_min)).or_default();
+
+        for idx in 0..data.num_rows() {
+            let max = ScalarValue::try_from_array(&maxs, idx)?;
+            let null_count = null_counts.values()[idx];
+            let offset = offsets.values()[idx];
+
+            map.entry(OrderableScalarValue(max))
+                .or_default()
+                .push(offset);
+            if null_count > 0 {
+                null_pages.push(offset);
+            }
+        }
+
+        // TODO: Support other page types?
+        let sub_index = Arc::new(FlatIndexLoader {});
+
+        Ok(Self::new(map, null_pages, store, sub_index))
+    }
+}
+
+#[async_trait]
+impl ScalarIndex for BTreeIndex {
+    async fn search(&self, query: &ScalarQuery) -> Result<UInt64Array> {
+        let pages = match query {
+            ScalarQuery::Equals(val) => self.page_lookup.pages_eq(val),
+            ScalarQuery::Range(start, end) => {
+                self.page_lookup.pages_between((start.clone(), end.clone()))
+            }
+            ScalarQuery::IsIn(values) => self.page_lookup.pages_in(values),
+            ScalarQuery::IsNull() => self.page_lookup.pages_null(),
+        };
+        let sub_index_reader = self.store.open_index_file(BTREE_PAGES_NAME).await?;
+        let page_tasks = pages
+            .into_iter()
+            .map(|page_index| {
+                self.search_page(query, page_index, sub_index_reader.clone())
+                    .boxed()
+            })
+            .collect::<Vec<_>>();
+        let row_id_lists = stream::iter(page_tasks)
+            .buffered(num_cpus::get())
+            .try_collect::<Vec<UInt64Array>>()
+            .await?;
+        let total_size = row_id_lists
+            .iter()
+            .map(|row_id_list| row_id_list.len())
+            .sum();
+        let mut all_row_ids = Vec::with_capacity(total_size);
+        for row_id_list in row_id_lists {
+            all_row_ids.extend(row_id_list.values());
+        }
+        Ok(UInt64Array::from_iter_values(all_row_ids))
+    }
+
+    async fn load(store: Arc<dyn IndexStore>) -> Result<Arc<Self>> {
+        let page_lookup_file = store.open_index_file(BTREE_LOOKUP_NAME).await?;
+        let serialized_lookup = page_lookup_file.read_record_batch(0).await?;
+        Ok(Arc::new(Self::try_from_serialized(
+            serialized_lookup,
+            store,
+        )?))
+    }
+}
+
+struct BatchStats {
+    min: ScalarValue,
+    max: ScalarValue,
+    null_count: u32,
+}
+
+fn min_val(array: &Arc<dyn Array>) -> Result<ScalarValue> {
+    let mut acc = MinAccumulator::try_new(array.data_type())?;
```

Comment:
Do we know the behavior for `NaN`, `Inf`, `-Inf` for floats?

Reply:
Hmmm, this is what DF uses internally for aggregate min.  This uses both [arrow::compute::kernels::aggregate::min](https://docs.rs/arrow/latest/arrow/compute/kernels/aggregate/fn.min.html#) which states:

> Returns the minimum value in the array, according to the natural order. For floating point arrays any NaN values are considered to be greater than any other non-null value

and also the `PartialOrd` implementation of `ScalarValue` which uses `total_cmp` for floats which puts negative `NaN` before all values which seems to contrast with the above statement :(.  I'll try some experiments.

Reply:
Turns out it is even weirder because there is a **third** interpretation of nan comparisons.  The `MinAccumulator` uses `f32::min` defined as "NaN is always larger" and `f32::max` defined as "NaN is always smaller" :laughing: 

So, DF has an issue (I'll file a ticket) but the btree itself is technically consistent.  Unfortunately, once we add training and create index then I think we have a problem because we will want the btree to be consistent with the training source (which should be ordered by val).  I'm pretty sure that relies on `arrow::compute::kernels::sort` which, from some testing, does seem to be using total order.

Probably the best fix is for DF to use total order for everything.  In the meantime we could perhaps raise a "not implemented" error if the min or max are ever NaN.

Reply:
I added a check to return an error if the column has NaN as the min/max for any page.